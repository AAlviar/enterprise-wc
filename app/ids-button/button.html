<ids-layout-grid>
  <ids-label font-size="12" type="h1">Buttons</ids-label>
</ids-layout-grid>
<ids-layout-grid>
  <ids-layout-column>
    <ids-button id="test-button-default">
      <span slot="text">Default Button</span>
    </ids-button>
  </ids-layout-column>

  <ids-layout-column>
    <ids-button id="test-button-primary" type="primary">
      <span slot="text">Primary Button</span>
    </ids-button>
  </ids-layout-column>

  <ids-layout-column>
    <ids-button id="test-button-secondary" type="secondary">
      <span slot="text">Secondary Button</span>
    </ids-button>
  </ids-layout-column>

  <ids-layout-column>
    <ids-button id="test-button-tertiary" type="tertiary">
      <span slot="text">Tertiary Button</span>
    </ids-button>
  </ids-layout-column>
</ids-layout-grid>
