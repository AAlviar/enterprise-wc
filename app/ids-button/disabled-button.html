<ids-layout-grid>
  <ids-label font-size="12" type="h1">Disabled Buttons</ids-label>
</ids-layout-grid>
<ids-layout-grid>
  <ids-layout-column>
    <ids-button id="test-disabled-button-default" disabled="true">
      <span slot="text">Default Button</span>
    </ids-button>
  </ids-layout-column>

  <ids-layout-column>
    <ids-button id="test-disabled-button-primary" type="primary"  disabled="true">
      <span slot="text">Primary Button</span>
    </ids-button>
  </ids-layout-column>

  <ids-layout-column>
    <ids-button id="test-disabled-button-secondary" type="secondary"  disabled="true">
      <span slot="text">Secondary Button</span>
    </ids-button>
  </ids-layout-column>

  <ids-layout-column>
    <ids-button id="test-disabled-button-tertiary" type="tertiary"  disabled="true">
      <span slot="text">Tertiary Button</span>
    </ids-button>
  </ids-layout-column>
</ids-layout-grid>
