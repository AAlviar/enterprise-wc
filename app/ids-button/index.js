import IdsButton from '../../src/ids-button/ids-button';

// Supporting components
import IdsIcon from '../../src/ids-icon/ids-icon';
import IdsLabel from '../../src/ids-label/ids-label';
import IdsLayoutGrid from '../../src/ids-layout-grid/ids-layout-grid';
