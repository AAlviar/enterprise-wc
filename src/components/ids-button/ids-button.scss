/**
 * Ids Button Component CSS
 * (Includes standard and icon button rules)
 */
@use '../../themes/mixins/ids-core-mixins' as mixins;
@import './ids-button-base';

:host {
  display: inline-flex;

  ::slotted(span) {
    vertical-align: bottom;
  }

  ::slotted(span),
  ::slotted(ids-text) {
    color: currentColor;
  }

  ::slotted(ids-icon) {
    vertical-align: middle;
    color: currentColor;
  }

  ::slotted(span.audible) {
    @include mixins.audible();
  }

  // Button elements should fill the space inside shadow root
  button,
  .ids-icon-button {
    justify-content: center;
    height: 100%;
    width: 100%;
  }
}

:host(.hidden) {
  display: none;
}

:host([hidden]) {
  display: none;
}

// Reset button display styles internally
button {
  display: block;
}

// Applies to all non-icon button types
.ids-button,
.ids-toggle-button,
.ids-menu-button:not(.ids-icon-button),
.ids-modal-button {
  @include mixins.antialiased();

  .ids-icon {
    display: inline-flex;
    place-self: center;
    vertical-align: middle;
  }
}

.ids-button,
.ids-toggle-button,
.ids-menu-button {
  @include ids-base-button-styles();
  @include ids-standard-button-colors();

  // Technically not valid
  ::slotted(ids-icon:only-child) {
    display: block;
  }

  // Assign padding based on type/usage
  // ====================================================
  &:not(.no-padding) {
    @include ids-standard-button-padding();
  }

  &.btn-tertiary,
  &.btn-tertiary-destructive {
    @include ids-inline-button-padding();
  }

  &.no-padding {
    @include ids-inline-button-padding();
  }
}

.ids-menu-button.formatter-width {
  &:not(.no-padding):not(.ids-icon-button) {
    padding-inline-start: var(--ids-spacing-lg);
    padding-inline-end: var(--ids-spacing-lg);
  }

  &:not(.formatter-width-percentage) {
    justify-content: normal;

    .ids-icon {
      position: absolute;
      inset-inline-end: var(--ids-spacing-md);
    }
  }
}

.ids-icon-button {
  @include ids-base-button-styles();
  @include ids-standard-button-colors();

  // Everything below is only for icon buttons
  padding-inline-start: var(--ids-spacing-lg);
  padding-inline-end: var(--ids-spacing-lg);
  padding-block-start: var(--ids-spacing-lg);
  padding-block-end: var(--ids-spacing-lg);

  &:not(.no-padding) {
    // 34px / 34px
    padding-inline-start: calc(var(--ids-spacing-lg) - 1px);
    padding-inline-end: calc(var(--ids-spacing-lg) - 1px);
    padding-top: calc(var(--ids-spacing-lg) - 1px);
    padding-bottom: calc(var(--ids-spacing-lg) - 1px);
  }

  &:not(.square) {
    border-radius: var(--ids-border-radius-md);
  }

  // Fixes an issue that causes the ripple effect not to appear "round" on Safari.
  // NOTE: Needs to be applied only when not focused because it cuts off the focus state.
  &:not(:focus) {
    mask-image: radial-gradient(var(--ids-button-tertiary-ripple-color), var(--ids-button-tertiary-ripple-color));
    /* stylelint-disable property-no-vendor-prefix */
    -webkit-mask-image: radial-gradient(var(--ids-button-tertiary-ripple-color), var(--ids-button-tertiary-ripple-color));
    /* stylelint-enable property-no-vendor-prefix */
  }

  .ids-icon {
    display: block;
    padding-inline-start: var(--ids-spacing-none);
    font-size: 0;
  }

  span.audible {
    @include mixins.audible();
  }

  .ripple-effect {
    height: var(--ids-size-2xl);
    width: var(--ids-size-2xl);
  }

  &.no-margins {
    margin-inline-start: var(--ids-spacing-none);
    margin-inline-end: var(--ids-spacing-none);
  }
}

// Mix-in special rules for contextual buttons
.ids-button,
.ids-icon-button,
.ids-menu-button {
  @include ids-standard-button-shadows();
}

// Specific Modifications for Modal Buttons
// Do not apply styles to modal close icon button, appearance as icon is required
.ids-modal-button:not(.ids-modal-icon-button) {
  @include ids-base-button-styles();
  @include ids-modal-button-colors();

  padding-block-start: var(--ids-spacing-xlg);
  padding-block-end: var(--ids-spacing-xlg);
  display: block;
  font-weight: var(--ids-font-weight-semi-bold);
  justify-content: center;
  height: 100%;
  width: 100%;

  &:focus {
    z-index: 1;
  }
}

// Adjusting hitbox for the modal close icon
.ids-modal-icon-button {
  padding: var(--ids-spacing-lg);
}

// ====================================================
// Swipe Action Buttons (Swipe action component)
.btn-swipe-action-left,
.btn-swipe-action-right {
  background-color: var(--ids-body-line-color);
  color: var(--ids-component-text-color);
  border-radius: 0;
  display: block;
  padding: 0 var(--ids-spacing-lg) var(--ids-spacing-lg);
  height: 100%;
  margin: 0;
  white-space: inherit;
  width: 85px;

  ::slotted(ids-icon) {
    display: block;
  }

  &:focus {
    box-shadow: none;
    border-color: transparent;
  }

  &:active {
    background-color: var(--ids-component-hover-color);
  }

  &:hover {
    background-color: var(--ids-component-hover-color);
    box-shadow: none;
    border-color: transparent;
  }
}

.btn-swipe-action-right {
  background-color: var(--ids-color-primary);
  color: var(--ids-inverse-text-color);

  &:hover {
    background-color: var(--ids-color-secondary);
    color: var(--ids-inverse-text-color);
  }

  &:active {
    background-color: var(--ids-color-secondary);
    color: var(--ids-inverse-text-color);
  }
}

// Applies to all non-icon button types
.ids-button,
.ids-toggle-button,
.ids-menu-button:not(.ids-icon-button),
.ids-modal-button:not(.ids-icon-button) {
  // Padding rules for buttons that insert both text/icon into the fallback (unnamed) slot.
  ::slotted(ids-icon:not(:first-child)), // shadowroot + text + icon
  ::slotted(span:nth-child(2):last-child), // shadowroot + icon + text
  ::slotted(ids-text:nth-child(2):last-child),
  ::slotted(span:nth-child(3):last-child), // shadowroot + icon + (audible text) + text
  ::slotted(ids-text:nth-child(3):last-child) {
    padding-inline-start: 5px;
  }

  // Padding rules for standalone CSS buttons (match the rules/style above)
  .ids-icon:last-child,
  span + span,
  span:last-child {
    padding-inline-start: 5px;
  }
}

// Adjustment for about close icon
.ids-modal-button {
  ::slotted(ids-icon:not(:first-child)) { // shadowroot + text + icon
    padding: var(--ids-spacing-sm);
  }
}

// Loading Indicator
:host(:not([show-loading-indicator='true'])) {
  ::slotted(.slot-loading-indicator) {
    display: none;
  }
}

:host([show-loading-indicator='true']) {
  ::slotted(ids-loading-indicator) {
    margin-top: 3px;
    margin-bottom: 3px;
  }
}

:host([show-loading-indicator='true'][loading-indicator-only='true']) {
  ::slotted(*:not(.slot-loading-indicator)) {
    visibility: hidden;
  }
}

:host([show-loading-indicator='true']:not([loading-indicator-only='true'])) {
  ::slotted(.slot-loading-indicator) {
    margin-inline-start: 6px;
  }
}

:host([show-loading-indicator='true']:not([loading-indicator-only='true']):is([type='primary'], [type='secondary'], [type='primary-destructive'])) {
  &::part(button) {
    padding-inline-start: 20px;
    padding-inline-end: 12px;
  }

  ::slotted(.slot-loading-indicator) {
    margin-inline-start: 10px;
  }
}

:host([loading-indicator-only='true']) {
  ::slotted(.slot-loading-indicator) {
    position: absolute;
    margin-inline-start: auto;
    margin-inline-end: auto;
    left: 0;
    right: 0;
  }
}
