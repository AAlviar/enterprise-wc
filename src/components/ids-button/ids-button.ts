import { customElement, scss } from '../../core/ids-decorators';
import { attributes } from '../../core/ids-attributes';
import { stringToBool } from '../../utils/ids-string-utils/ids-string-utils';

import Base from './ids-button-base';
import {
  BUTTON_TYPES, BUTTON_DEFAULTS, BUTTON_ATTRIBUTES, ICON_ALIGN, baseProtoClasses
} from './ids-button-attributes';

import styles from './ids-button.scss';
import IdsIcon from '../ids-icon/ids-icon';
import IdsText from '../ids-text/ids-text';

/**
 * IDS Button Component
 * @type {IdsButton}
 * @inherits IdsElement
 * @mixes IdsEventsMixin
 * @mixes IdsColorVariantMixin
 * @mixes IdsLocaleMixin
 * @mixes IdsRippleMixin
 * @mixes IdsThemeMixin
 * @mixes IdsTooltipMixin
 * @part button - the button element
 * @part icon - the icon element
 * @part text - the text element
 */
@customElement('ids-button')
@scss(styles)
export default class IdsButton extends Base {
  shouldUpdate = true;

  constructor() {
    super();
    Object.keys(BUTTON_DEFAULTS).forEach((prop) => {
      this.state[prop] = BUTTON_DEFAULTS[prop];
    });
  }

  /**
   * Button-level `connectedCallback` implementation
   * @returns {void}
   */
  connectedCallback(): void {
    super.connectedCallback();
    this.#setInitialState();
  }

  #setInitialState() {
    if (this.hasAttribute(attributes.ICON)) this.appendIcon(this.getAttribute(attributes.ICON));
    if (this.hasAttribute(attributes.TEXT)) this.appendText(this.getAttribute(attributes.TEXT));

    const isIconButton = this.button?.classList.contains('ids-icon-button');
    this.setupRipple(this.button as HTMLButtonElement, isIconButton ? 35 : 50);
    this.setIconAlignment();
    this.refreshProtoClasses();
  }

  /**
   * Return the attributes we handle as getters/setters
   * @returns {Array<string>} The attributes in an array
   */
  static get attributes(): Array<string> {
    return [...super.attributes, ...BUTTON_ATTRIBUTES];
  }

  /**
   * Inherited from `IdsColorVariantMixin`
   * @returns {Array<string>} List of available color variants for this component
   */
  colorVariants = ['alternate', 'alternate-formatter'];

  /**
   * Figure out the classes
   * @private
   * @readonly
   * @returns {Array<string>} containing classes used to identify this button prototype
   */
  get protoClasses() {
    const textSlot = this.querySelector('span:not(.audible), ids-text:not([audible])');
    const iconSlot = this.querySelector('ids-icon[slot]') || this.querySelector('ids-icon');
    if (iconSlot && (!textSlot)) {
      return ['ids-icon-button'];
    }
    return ['ids-button'];
  }

  /**
   * Refreshes this button's prototype CSS class
   * @private
   * @returns {void}
   */
  refreshProtoClasses() {
    if (!this.button) return;
    const cl = this.button.classList;
    const newProtoClass = this.protoClasses;

    cl.remove(...baseProtoClasses);
    cl.add(...newProtoClass);
  }

  /**
   * Inner template contents
   * @returns {string} The template
   */
  template() {
    let cssClass = '';
    let protoClasses = '';
    let disabled = '';
    let tabIndex = 'tabindex="0"';
    let type = '';
    if (this.state?.cssClass) {
      cssClass = ` ${this.state.cssClass.join(' ')}`;
    }
    if (this.state?.disabled) {
      disabled = ` disabled="true"`;
    }
    if (this.state?.tabIndex) {
      tabIndex = `tabindex="${this.state.tabIndex}"`;
    }
    if (this.state && this.state?.type !== 'default') {
      type = ` btn-${this.state.type}`;
    }

    if (this.hasAttribute(attributes.SQUARE)) {
      cssClass += ' square';
    }

    if (this.hasAttribute(attributes.NO_PADDING)) {
      cssClass += ' no-padding';
    }

    if (this.protoClasses.length) {
      protoClasses = `${this.protoClasses.join(' ')}`;
    }

    const alignCSS = this.state?.iconAlign === 'end'
      ? ' align-icon-end'
      : ' align-icon-start';

    const slots = this.#templateSlots();

    return `<button part="button" class="${protoClasses}${type}${alignCSS}${cssClass}" ${tabIndex}${disabled}>
      ${slots}
    </button>`;
  }

  #templateSlots() {
    const namedSlots = this.state?.iconAlign === 'end'
      ? `<slot name="text" part="text"></slot><slot name="icon"></slot>`
      : `<slot name="icon" part="icon"></slot><slot name="text"></slot>`;
    return `${namedSlots}<slot></slot>`;
  }

  /**
   * @readonly
   * @returns {HTMLButtonElement|undefined|null} reference to the true button element used in the Shadow Root
   */
  get button(): HTMLButtonElement | undefined | null {
    return this.shadowRoot?.querySelector<HTMLButtonElement>('button');
  }

  /**
   * @param {Array<string>|string} val containing CSS classes that will be applied to the button
   * Strings will be split into an array and separated by whitespace.
   */
  set cssClass(val) {
    let attr = val;
    let newCl: any[] = [];
    // @TODO replace with clone utils method
    const prevClasses = [].concat(this.state.cssClass);

    if (Array.isArray(val)) {
      newCl = val;
      attr = val.join(' ');
    } else if (typeof val === 'string' && val.length) {
      newCl = val.split(' ');
    }

    this.state.cssClass = newCl;
    if (newCl.length) {
      this.setAttribute(attributes.CSS_CLASS, attr.toString());
    } else {
      this.removeAttribute(attributes.CSS_CLASS);
    }

    // Remove/Set CSS classes on the actual inner Button component
    if (!this.button) return;

    const buttonCl = this.button.classList;
    const buttonClArr = Array.from(buttonCl);
    prevClasses.forEach((cssClass) => {
      if (!newCl.includes(cssClass)) {
        buttonCl.remove(cssClass);
      }
    });
    newCl.forEach((newCssClass) => {
      if (!buttonClArr.includes(newCssClass)) {
        buttonCl.add(newCssClass);
      }
    });
  }

  get cssClass(): string {
    return this.state.cssClass;
  }

  /**
   * Passes a disabled attribute from the custom element to the button
   * @param {boolean|string} val true if the button will be disabled
   */
  set disabled(val: boolean | string) {
    const isValueTruthy = stringToBool(val);
    this.shouldUpdate = false;

    if (isValueTruthy) {
      this.setAttribute(attributes.DISABLED, '');
    } else {
      this.removeAttribute(attributes.DISABLED);
    }

    this.shouldUpdate = true;
    this.state.disabled = isValueTruthy;
    if (this.button) this.button.disabled = isValueTruthy;
  }

  get disabled(): boolean {
    return this.state.disabled;
  }

  /**
   * @param {boolean|string} val true if the button component should be hidden from view
   */
  set hidden(val: boolean | string) {
    const isValueTruthy = stringToBool(val);
    this.shouldUpdate = false;
    if (isValueTruthy) {
      this.setAttribute(attributes.HIDDEN, '');
    } else {
      this.removeAttribute(attributes.HIDDEN);
    }

    this.shouldUpdate = true;
    this.state.hidden = isValueTruthy;
    if (this.button) this.button.hidden = isValueTruthy;
  }

  /**
   * @returns {boolean} true if the button component is hidden from view
   */
  get hidden(): boolean {
    return this.state.hidden;
  }

  /**
   * Passes a tabIndex attribute from the custom element to the button
   * @param {number | string | null} val the tabIndex value
   */
  set tabIndex(val: number | string | null) {
    const trueVal = Number(val);

    this.removeAttribute(attributes.TABINDEX);

    if (Number.isNaN(trueVal) || trueVal < -1) {
      this.state.tabIndex = 0;
      this.button?.setAttribute(attributes.TABINDEX, '0');
      return;
    }

    this.state.tabIndex = trueVal;
    this.button?.setAttribute(attributes.TABINDEX, `${trueVal}`);
  }

  /**
   * @returns {number} the current tabIndex number for the button
   */
  get tabIndex(): number {
    return this.state.tabIndex;
  }

  /**
   * Sets the icon on the button
   * @param {string | undefined} val representing the icon to set
   */
  set icon(val) {
    if (typeof val !== 'string' || !val.length) {
      this.removeAttribute(attributes.ICON);
      this.state.icon = undefined;
      this.removeIcon();
      return;
    }
    this.state.icon = val;
    this.setAttribute(attributes.ICON, val);
    this.appendIcon(val);
  }

  /**
   * Gets the current icon used on the button
   * @returns {string | null} a defined IdsIcon's `icon` attribute, if one is present
   */
  get icon(): string | null {
    return this.querySelector('ids-icon')?.getAttribute('icon') || null;
  }

  /**
   * Gets the current icon element
   * @readonly
   * @returns {HTMLElement | null} a defined IdsIcon, if one is present
   */
  get iconEl(): HTMLElement | null {
    return this.querySelector('ids-icon');
  }

  /**
   * Sets the alignment of an existing icon to the 'start' or 'end' of the text
   * @param {string} val the alignment type to set.
   */
  set iconAlign(val: string) {
    if (!ICON_ALIGN.includes(`align-icon-${val}`)) {
      val = 'start';
    }
    this.state.iconAlign = val;
    this.setIconAlignment();
  }

  /**
   * @returns {string} containing 'start' or 'end'
   */
  get iconAlign(): string {
    return this.state?.iconAlign || 'start';
  }

  /**
   * Get width
   * @returns {string | null} 100%, 90px, 50rem etc.
   */
  get width(): string | null {
    return this.getAttribute('width');
  }

  /**
   * Set width of button
   * @param {string | null} w 100%, 90px, 50rem etc.
   */
  set width(w: string | null) {
    if (!w) {
      this.removeAttribute('width');
      this.style.width = '';
      if (this.button) this.button.style.width = '';
      return;
    }

    // if percentage passed set width to host
    if (w.indexOf('%') !== -1) {
      this.style.width = w;
      if (this.button) this.button.style.width = '';
    } else {
      this.style.width = '';
      if (this.button) this.button.style.width = w;
    }

    this.setAttribute('width', w);
  }

  /**
   * Check if an icon exists, and adds the icon if it's missing
   * @param {string | null} iconName The icon name to check
   * @private
   */
  appendIcon(iconName: string) {
    const icon = this.querySelector<IdsIcon>(`ids-icon`); // @TODO check for dropdown/expander icons here

    if (icon) {
      icon.icon = iconName;
      this.setIconAlignment();
    } else {
      this.insertAdjacentHTML('afterbegin', `<ids-icon slot="icon" icon="${iconName}" class="ids-icon"></ids-icon>`);
    }

    this.refreshProtoClasses();
  }

  /**
   * Check if an icon exists, and removes the icon if it's present
   * @private
   */
  removeIcon(): void {
    const icon = this.querySelector(`ids-icon`); // @TODO check for dropdown/expander icons here

    if (icon) {
      icon.remove();
    }
    this.setIconAlignment();
    this.refreshProtoClasses();
  }

  /**
   * Adds/Removes Icon Alignment CSS classes to/from the inner button component.
   * @private
   */
  setIconAlignment(): void {
    if (!this.button) return;

    const alignment = this.iconAlign || 'start';
    const iconStr = this.icon;
    this.button.classList.remove(...ICON_ALIGN);

    // Append the icon, if needed
    if (iconStr) {
      this.button.classList.add(`align-icon-${alignment}`);
    }

    // Re-arrange the slots
    const iconSlot = this.button.querySelector('slot[name="icon"]');
    if (!iconSlot) {
      return;
    }

    if (alignment === 'end') {
      this.button.appendChild(iconSlot);
    } else {
      this.button.prepend(iconSlot);
    }
  }

  /**
   * @param {string | null} val the text value
   * @returns {void}
   */
  set text(val: string | null) {
    this.removeAttribute(attributes.TEXT);

    if (typeof val !== 'string' || !val.length) {
      this.state.text = '';
      this.removeText();
      return;
    }

    // @TODO: Run this through an XSS check
    this.state.text = val;
    this.appendText(val);
  }

  /**
   * @returns {string | null} the current text value
   */
  get text(): string | null {
    const textElem = this.querySelector('span:not(.audible)');
    if (textElem && textElem.textContent?.length) {
      return textElem.textContent;
    }
    return this.textContent;
  }

  /**
   * Check if the text slot exists, and appends it if it's missing
   * @param {string|null} val New text contents
   * @private
   */
  appendText(val: string | null) {
    const text = this.querySelector(`span:not(.audible)`);

    if (text) {
      text.textContent = val;
    } else {
      this.insertAdjacentHTML('afterbegin', `<span>${val}</span>`);
    }

    this.refreshProtoClasses();
  }

  /**
   * Checks if the text slot exists, and removes it if necessary
   * @private
   */
  removeText() {
    const text = this.querySelector(`span:not(.audible)`);
    if (text) {
      text.remove();
    }
    this.refreshProtoClasses();
  }

  /**
   * Set the button types between 'default', 'primary', 'secondary', 'tertiary', or 'destructive'
   * @param {string | null} val a valid button "type"
   */
  set type(val: string | null) {
    if (!val || BUTTON_TYPES.indexOf(val) <= 0) {
      this.removeAttribute(attributes.TYPE);
      this.state.type = BUTTON_TYPES[0];
    } else {
      this.setAttribute(attributes.TYPE, val);
      if (this.state.type !== val) {
        this.state.type = val;
      }
    }
    this.setTypeClass(val);
  }

  /**
   * @returns {string} the currently set type
   */
  get type(): string {
    return this.state.type;
  }

  /**
   * Sets the no margins attribute
   * @param {boolean | string} n string value from the no margins attribute
   */
  set noMargins(n: boolean | string) {
    if (stringToBool(n)) {
      this.setAttribute(attributes.NO_MARGINS, '');
      this.container?.classList.add(attributes.NO_MARGINS);
      return;
    }
    this.removeAttribute(attributes.NO_MARGINS);
    this.container?.classList.remove(attributes.NO_MARGINS);
  }

  get noMargins(): boolean {
    return stringToBool(this.getAttribute(attributes.NO_MARGINS));
  }

  /**
   * @param {boolean | string} val true if the button should not have standard padding rules applied
   */
  set noPadding(val: boolean | string) {
    const isTruthy = this.noPadding;
    const trueVal = stringToBool(val);
    if (isTruthy !== trueVal) {
      if (trueVal) {
        this.container?.classList.add('no-padding');
        this.setAttribute('no-padding', 'true');
      } else {
        this.container?.classList.remove('no-padding');
        this.removeAttribute('no-padding');
      }
    }
  }

  /**
   * @returns {boolean | string} true if the button does not currently have standard padding rules applied
   */
  get noPadding(): boolean | string {
    return stringToBool(this.getAttribute(attributes.NO_PADDING)); // this.container.classList.contains('no-padding');
  }

  /**
   * @param {boolean} value whether the corners of the button as an icon-button should be angled/90°
   */
  set square(value: boolean) {
    const isTruthy = stringToBool(value);

    if (this.button) {
      if (isTruthy && !this.button.classList.contains('square')) {
        this.button.classList.add('square');
      } else if (!isTruthy && this.button.classList.contains('square')) {
        this.button.classList.remove('square');
      }
    }

    if (isTruthy && !this.hasAttribute(attributes.SQUARE)) {
      this.setAttribute(attributes.SQUARE, '');
    } else if (!isTruthy && this.hasAttribute(attributes.SQUARE)) {
      this.removeAttribute(attributes.SQUARE);
    }
  }

  /**
   * @returns {boolean} whether the corners of the button as an icon-button are angled/90°
   */
  get square(): boolean {
    return this.hasAttribute(attributes.SQUARE);
  }

  /**
   * Sets the correct type class on the Shadow button.
   * @private
   * @param {string | null} val desired type class
   */
  setTypeClass(val: string | null) {
    if (this.button) {
      BUTTON_TYPES.forEach((type) => {
        const typeClassName = `btn-${type}`;
        if (val === type) {
          if (type !== 'default' && !this.button?.classList.contains(typeClassName)) {
            this.button?.classList.add(typeClassName);
          }
          return;
        }
        if (this.button?.classList.contains(typeClassName)) {
          this.button?.classList.remove(typeClassName);
        }
      });
    }
  }

  /**
   * Overrides the standard "focus" behavior to instead pass focus to the inner HTMLButton element.
   */
  focus(): void {
    this.button?.focus();
  }

  /**
   * Implements callback from IdsColorVariantMixin used to
   * update the color variant on children components
   * @returns {void}
   */
  onColorVariantRefresh(): void {
    const icons = this.querySelectorAll<IdsIcon>('ids-icon');
    const texts = this.querySelectorAll<IdsText>('ids-text');

    const iterator = (el: IdsIcon | IdsText) => {
      el.colorVariant = this.colorVariant;
    };

    [...icons, ...texts].forEach(iterator);
  }
}
