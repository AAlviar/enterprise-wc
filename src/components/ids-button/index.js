export { IdsButton, BUTTON_ATTRIBUTES, BUTTON_TYPES } from './ids-button';
