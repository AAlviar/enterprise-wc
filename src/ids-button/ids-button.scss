/**
 * Ids Button Component CSS
 * (Includes standard and icon button rules)
 */
@import './ids-button-base';

:host {
  @include inline-flex();
}

// Standard slot rules
::slotted(ids-icon),
::slotted(span) {
  vertical-align: middle;
}

// Icon button slot rules
::slotted(span.audible) {
  @include audible();

  vertical-align: middle;
}

::slotted(ids-icon:only-child) {
  @include block();
}

.ripple-effect {
  @include block();
  @include rounded-round();

  opacity: 0;
  position: absolute;
  transform: scale(0);

  &.animating {
    @include animation(ripple-animation 1.2s ease-out);
  }
}

.ids-button {
  @include ids-base-button-styles();

  .ripple-effect {
    height: 250px;
    width: 250px;
  }
}

.ids-icon-button {
  @include ids-base-button-styles();

  // Everything below is only for icon buttons
  @include px-8();
  @include py-8();
  @include rounded-round();

  .ripple-effect {
    height: 70px;
    width: 70px;
  }
}

@include keyframes(ripple-animation) {
  from {
    @include opacity-50();

    transform: scale(0);
  }

  to {
    @include opacity-0();

    transform: scale(2);
  }
}
