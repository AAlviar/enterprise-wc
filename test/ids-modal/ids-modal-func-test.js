/**
 * @jest-environment jsdom
 */
import IdsModal, { IdsOverlay } from '../../src/ids-modal';
import { IdsButton } from '../../src/ids-button/ids-button';

describe('IdsModal Component', () => {
  let modal;

  beforeEach(async () => {
    const elem = new IdsModal();
    document.body.appendChild(elem);
    modal = document.querySelector('ids-modal');
  });

  afterEach(async () => {
    document.body.innerHTML = '';
  });

  it('renders with no errors', () => {
    const errors = jest.spyOn(global.console, 'error');
    const elem = new IdsModal();
    document.body.appendChild(elem);
    elem.remove();
    expect(document.querySelectorAll('ids-modal').length).toEqual(1);
    expect(errors).not.toHaveBeenCalled();
  });

  it('renders correctly', () => {
    expect(modal.outerHTML).toMatchSnapshot();
    modal.show();
    expect(modal.outerHTML).toMatchSnapshot();

    modal.hide();
    expect(modal.outerHTML).toMatchSnapshot();
  });

  it('can show/hide by using attributes', () => {
    modal.setAttribute('visible', 'true');

    expect(modal.visible).toBeTruthy();

    modal.removeAttribute('visible');

    expect(modal.visible).toBeFalsy();
  });

  it('can show/hide by using attributes', () => {
    modal.visible = true;

    expect(modal.visible).toBeTruthy();

    modal.visible = false;

    expect(modal.visible).toBeFalsy();
  });

  it('can use `show()`/`hide()` methods', () => {
    modal.show();

    expect(modal.visible).toBeTruthy();

    modal.hide();

    expect(modal.visible).toBeFalsy();
  });

  it('can prevent being opened with the `beforeshow` event', () => {
    modal.addEventListener('beforeshow', (e) => {
      e.detail.response(false);
    });
    modal.show();

    expect(modal.visible).toBeFalsy();
  });

  it('can prevent being closed with the `beforehide` event', (done) => {
    modal.addEventListener('beforehide', (e) => {
      e.detail.response(false);
    });
    modal.show();

    setTimeout(() => {
      modal.hide();

      setTimeout(() => {
        expect(modal.visible).toBeTruthy();
        done();
      }, 70);
    }, 70);
  });

  it('can have a target element', () => {
    const btn = new IdsButton();
    btn.type = 'secondary';
    btn.text = 'I am the target';
    document.body.appendChild(btn);

    modal.target = btn;

    expect(modal.state.target.isEqualNode(btn)).toBeTruthy();

    // Clicking on the trigger should make the Modal show
    const clickEvent = new MouseEvent('click', { bubbles: true });
    btn.dispatchEvent(clickEvent);

    // Dispatch twice to cover the 'else'
    btn.dispatchEvent(clickEvent);

    modal.target = null;

    expect(modal.state.target).toBeNull();
  });

  it('can have a title', () => {
    modal.title = 'I have a title';

    expect(modal.state.title).toEqual('I have a title');
    expect(modal.querySelectorAll('[slot="title"]').length).toBe(1);

    // Removes the slotted element if no title is present
    modal.title = '';

    expect(modal.state.title).toEqual('');
    expect(modal.querySelectorAll('[slot="title"]').length).toBe(0);

    // Adds it back if we apply a new title
    modal.title = 'New title';

    expect(modal.state.title).toEqual('New title');
    expect(modal.querySelectorAll('[slot="title"]').length).toBe(1);
  });

  it('can use an external overlay, if applicable', () => {
    const overlay = new IdsOverlay();
    modal.overlay = overlay;

    expect(modal.state.overlay.isEqualNode(overlay)).toBeTruthy();

    modal.overlay = null;

    // Internal overlay state should be empty,
    // but the Modal will use its own overlay instance internally
    expect(modal.state.overlay).toBeNull();
    expect(modal.container.querySelector('ids-overlay')).toBeDefined();
  });

  // @TODO Not sure why this won't pass (structured similarly to other tests like it)
  it.skip('can hide a visible modal by pressing the Escape key', (done) => {
    const closeEvent = new KeyboardEvent('keydown', { key: 'Escape', bubbles: true });

    modal.show();

    setTimeout(() => {
      // Key event is captured within the Modal container (encapsulated)
      modal.container.dispatchEvent(closeEvent);
      setTimeout(() => {
        expect(modal.popup.visible).toBeFalsy();
        done();
      }, 70);
    }, 70);
  });

  it('will not trigger a vetoable event of any type not supported', () => {
    modal.triggerVetoableEvent('fish');

    expect(modal.state.visible).toBeFalsy();
  });

  it('will focus the first available item in the modal when it opens', (done) => {
    const extraBtn = new IdsButton();
    extraBtn.id = 'focusable';
    extraBtn.type = 'secondary';
    modal.appendChild(extraBtn);
    modal.show();

    setTimeout(() => {
      expect(document.activeElement.isEqualNode(extraBtn)).toBeTruthy();
      done();
    }, 300);
  });
});
