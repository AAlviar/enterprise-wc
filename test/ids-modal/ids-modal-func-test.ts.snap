// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdsModal Component renders correctly 1`] = `"<ids-modal role="dialog" fullsize="sm" respond-down="sm"></ids-modal>"`;

exports[`IdsModal Component renders correctly 2`] = `"<ids-modal role="dialog" fullsize="sm" respond-down="sm" visible="" style="z-index: 1020;"></ids-modal>"`;

exports[`IdsModal Component renders correctly 3`] = `"<ids-modal role="dialog" fullsize="sm" respond-down="sm" style="z-index: 1020;"></ids-modal>"`;
